prime-router/src/main/kotlin/azure/ReportFunction.kt

```diff
@@ -103,103 +109,140 @@ class ReportFunction : Logging {
-                }
-            }
-        } catch (e: Exception) {
-            context.logger.log(Level.SEVERE, e.message, e)
-            actionHistory.trackActionResult("Exception: ${e.message ?: e}")
-            HttpUtilities.internalErrorResponse(request)
+
+        try {
+            return processRequest(request, sender, context, workflowEngine, actionHistory)
+        } catch (ex: Exception) {
+            if (ex.message != null)
+                logger.error(ex.message!!, ex)
+            else
+                logger.error(ex)
+            return HttpUtilities.internalErrorResponse(request)
         }
-        actionHistory.trackActionResult(httpResponseMessage)
-        // add the response to the action table as JSONB and record the httpsStatus
-        actionHistory.trackActionResponse(httpResponseMessage, verboseResponse.toString())
-        workflowEngine.recordAction(actionHistory)
-        actionHistory.queueMessages(workflowEngine) // Must be done after creating TASK record.
+    }
 
+    private fun processRequest(
+        request: HttpRequestMessage<String?>,
+        sender: Sender,
+        context: ExecutionContext,
+        workflowEngine: WorkflowEngine,
+        actionHistory: ActionHistory
+    ): HttpResponseMessage {
+        // determine if we should be following the sync or async workflow
+        val isAsync = processingType(request) == ProcessingType.Async
+        val errors: MutableList<ResultDetail> = mutableListOf()
+        val warnings: MutableList<ResultDetail> = mutableListOf()
+        // The following is identical to waters (for arch reasons)
+        val validatedRequest = validateRequest(workflowEngine, request)
+        warnings += validatedRequest.warnings
+        handleValidation(validatedRequest, request, actionHistory, workflowEngine)?.let {
+            return it
+        }
+
+        var report = workflowEngine.createReport(
+            sender,
+            validatedRequest.content,
+            validatedRequest.defaults,
+            errors,
+            warnings
+        )
+
+        // if we are processing a message asynchronously, the report's next action will be 'process'
+        if (isAsync && report != null) {
+            report.nextAction = TaskAction.process
+        }
+
+        // checks for errors from createReport
+        if (validatedRequest.options != Options.SkipInvalidItems && errors.isNotEmpty()) {
+            val responseBody = actionHistory.createResponseBody(
+                validatedRequest.options,
+                warnings,
+                errors,
+                false
+            )
+            val response = HttpUtilities.httpResponse(
+                request,
+                responseBody,
+                HttpStatus.BAD_REQUEST
+            )
+            actionHistory.trackActionResult(response)
+            actionHistory.trackActionResponse(response, responseBody)
+            workflowEngine.recordAction(actionHistory)
+            return response
```

Comment:
Try to limit the number of returns you do in a function.  Aim for just one if you can, but many tends to decrease readability.  Consider using an else here.

Reply:
In this case, 'else' makes sense. Updated.